test/e2e/upgrade_test.go

```diff
@@ -0,0 +1,146 @@
+	require.NoError(t, err)
+
+	for _, node := range nodes.Items {
+		patchData := fmt.Sprintf(`{"metadata":{"annotations":{"%s":"%s"}}}`, nc.VersionAnnotation, "badVersion")
+		_, err := testCtx.kubeclient.CoreV1().Nodes().Patch(context.TODO(), node.Name, types.MergePatchType, []byte(patchData), metav1.PatchOptions{})
+		require.NoError(t, err)
+		if err == nil {
+			break
+		}
+	}
+
+	err = testCtx.waitForWindowsNodes(gc.numberOfNodes, true, false, true)
+	require.NoError(t, err, "windows node upgrade failed")
+	// Test if the version annotation corresponds to the current operator version
+	testVersionAnnotation(t)
+}
+
+// configureUpgradeTest carries out steps required before running tests for upgrade scenario.
+// The steps include -
+// 1. Scale down the operator to 0.
+// 2. Change Windows node version annotation to an invalid value
+// 3. Scale up the operator to 1
+func (tc *testContext) configureUpgradeTest() error {
+	// Scale down the WMCO deployment to 0
+	if err := tc.scaleWMCODeployment(0); err != nil {
+		return err
+	}
+
+	// Override the Windows Node Version Annotation
+	nodes, err := tc.kubeclient.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{LabelSelector: nc.WindowsOSLabel})
+	if err != nil {
+		return err
+	}
+	if len(nodes.Items) != int(gc.numberOfNodes) {
+		return errors.Wrapf(nil, "unexpected number of nodes %v", gc.numberOfNodes)
+	}
+
+	for _, node := range nodes.Items {
+		patchData := fmt.Sprintf(`{"metadata":{"annotations":{"%s":"%s"}}}`, nc.VersionAnnotation, "badVersion")
+		_, err := tc.kubeclient.CoreV1().Nodes().Patch(context.TODO(), node.Name, types.MergePatchType, []byte(patchData), metav1.PatchOptions{})
+		if err != nil {
+			return err
+		}
+		log.Printf("Node Annotation changed to %v", node.Annotations[nc.VersionAnnotation])
+	}
+
+	// Scale up the WMCO deployment to 1
+	if err := tc.scaleWMCODeployment(1); err != nil {
+		return err
+	}
+	return nil
+}
+
+// scaleWMCODeployment scales the WMCO operator to the given replicas. If the deployment is managed by OLM, updating the
+// replicas only scales the deployment to 0 or 1. If we want to scale the deployment to more than 1 replicas, we need to
+// make changes in replicas defined in the corresponding CSV.
+func (tc *testContext) scaleWMCODeployment(desiredReplicas int32) error {
+	// update the windows-machine-config-operator deployment to the desired replicas - 0 or 1
+	err := wait.Poll(deploymentRetryInterval, deploymentTimeout, func() (done bool, err error) {
+
+		patchData := fmt.Sprintf(`{"spec":{"replicas":%v}}`, desiredReplicas)
+
+		_, err = tc.kubeclient.AppsV1().Deployments(resourceName).Patch(context.TODO(), resourceName,
+			types.MergePatchType, []byte(patchData), metav1.PatchOptions{})
+		if err != nil {
+			log.Printf("error patching operator deployment : %v", err)
+			return false, nil
+		}
+		return true, nil
+	})
+
+	// wait for the windows-machine-config-operator to scale up/down
+	if err == nil {
```

Comment:
Instead of this, you should do
```
if err != nil {
  return ...
}
err = wait.Poll(...
```
This is a better way of doing this as
1) it reduces indentation. The rest of the function is indented which is something that is nice to avoid
2) the control flow is more obvious. If we get an error we want to return immediately. If this function was to be added to in the future things would get pretty messy.